### ISSUE TITLE
[numpy] mypy does not raise `type-var` when `NDArray` with incompatible dtype is used as a generic.

### ISSUE THREAD

**Issue by randolf-scholz** [python/mypy#17228]
**To Reproduce**

```python
import numpy as np
from numpy.typing import NDArray

class A: ...

def foo() -> NDArray[A]:  # ✅ raises type-var
    return np.array([A()])

def bar() -> list[NDArray[A]]:  # ❌ does not raise type-var
    return [np.array([A()])]
```

**Expected Behavior**

`mypy` should raise `error: Type argument "A" of "NDArray" must be a subtype of "generic"  [type-var]` in both instances.

**Your Environment**

- Mypy version used: 1.10
- numpy version: 1.26.4
- Mypy configuration options: `plugins = ["numpy.typing.mypy_plugin"]`
- Python version used: 3.11.7
